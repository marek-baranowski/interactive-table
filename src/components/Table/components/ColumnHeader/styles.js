import { css } from "glamor";

export const headerTitleStyles = css({
  flex: 1,
  padding: "0 5px"
});

export const iconStyles = css({
  cursor: "pointer",
  fontWeight: "normal"
});
