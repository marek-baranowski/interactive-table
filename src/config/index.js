export const FILTER_TYPES = {
  STRING_FILTER: "StringFilter",
  MULTI_SELECT_FILTER: "MultiSelectFilter",
  RANGE_FILTER: "RangeFilter"
};

export const SORT_ORDER_TYPES = {
  ASCENDING: "asc",
  DESCENDING: "desc"
};

export const ASYNC_STATES = {
  NONE: "NONE",
  PENDING: "PENDING",
  RESOLVED: "RESOLVED",
  REJECTED: "REJECTED"
};

export const COLUMNS = [
  {
    key: "name",
    header: "Name",
    filter: { type: FILTER_TYPES.STRING_FILTER }
  },
  {
    key: "animal",
    header: "Animal",
    filter: { type: FILTER_TYPES.MULTI_SELECT_FILTER }
  },
  { key: "colour", header: "Colour" },
  { key: "pattern", header: "Pattern" },
  { key: "rating", header: "Rating", sortable: true },
  {
    key: "price",
    header: "Price",
    filter: { type: FILTER_TYPES.RANGE_FILTER },
    sortable: true
  }
];
