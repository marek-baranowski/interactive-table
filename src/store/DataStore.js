import { types, flow, getEnv } from "mobx-state-tree";
import orderBy from "lodash/orderBy";
import {
  StringFilterModel,
  MultiSelectFilterModel,
  RangeFilterModel
} from "./FilterModels";
import SortingModel from "./SortingModel";
import AsyncStatusModel from "./AsyncStatusModel";

const ColumnModel = types.model("ColumnModel", {
  key: types.string,
  header: types.string,
  filter: types.maybe(
    types.union(StringFilterModel, MultiSelectFilterModel, RangeFilterModel)
  ),
  sortable: types.maybe(types.boolean)
});

export default types
  .model("DataStore", {
    records: types.optional(types.array(types.frozen), []),
    columns: types.optional(types.array(ColumnModel), []),
    sorting: types.optional(SortingModel, {}),
    requestStatus: types.optional(AsyncStatusModel, {})
  })
  .views(self => ({
    get columnsWithFilter() {
      return self.columns.filter(column => !!column.filter);
    },
    get filteredSortedRecords() {
      const { records, columnsWithFilter, sorting } = self;
      const filtered = records.filter(record =>
        columnsWithFilter
          .map(({ filter }) => filter.validateRecord(record))
          .every(result => result)
      );

      return sorting.column
        ? orderBy(filtered, sorting.column, sorting.order)
        : filtered;
    }
  }))
  .actions(self => ({
    fetchRecords: flow(function*(service) {
      self.requestStatus.setPending();
      try {
        self.records = yield service();
        self.requestStatus.setResolved();
      } catch (error) {
        self.requestStatus.setRejected(error);
      }
    }),
    afterCreate: () => self.fetchRecords(getEnv(self).service)
  }));
