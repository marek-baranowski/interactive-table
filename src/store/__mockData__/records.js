export default [
  {
    name: "Guadalupe",
    animal: "Dog",
    colour: "White",
    pattern: "Spotted",
    rating: 8,
    price: 285
  },
  {
    name: "Caroline",
    animal: "Turtle",
    colour: "Black",
    pattern: "Striped",
    rating: 2,
    price: 911
  },
  {
    name: "Yvette",
    animal: "Turtle",
    colour: "Green",
    pattern: "Striped",
    rating: 4,
    price: 653
  },
  {
    name: "Hattie",
    animal: "Dog",
    colour: "White",
    pattern: "Solid",
    rating: 3,
    price: 738
  },
  {
    name: "Hattie",
    animal: "Turtle",
    colour: "Black",
    pattern: "Solid",
    rating: 5,
    price: 548
  },
  {
    name: "Hatson",
    animal: "Snake",
    colour: "White",
    pattern: "Spotted",
    rating: 7,
    price: 439
  }
];
