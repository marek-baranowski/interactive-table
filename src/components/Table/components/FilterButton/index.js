import React from "react";
import { Popover, PopoverBody } from "reactstrap";
import { observer } from "mobx-react";
import { iconStyles } from "../ColumnHeader/styles";
import StringFilter from "../StringFilter";
import MultiSelectFilter from "../MultiSelectFilter";
import RangeFilter from "../RangeFilter";
import { FILTER_TYPES } from "config";

class FilterButton extends React.Component {
  handleRef = element => {
    !!element && (this.popoverTarget = element);
  };

  render() {
    const { filter } = this.props;

    if (!filter) {
      return null;
    }

    const iconProps = {
      ref: this.handleRef,
      onClick: filter.toggleVisibility,
      className: `fa fa-filter ${iconStyles}${filter.isActive
        ? " text-primary"
        : ""}`
    };

    return (
      <span>
        <i {...iconProps} />
        <Popover
          placement="top"
          isOpen={filter.isVisible}
          target={() => this.popoverTarget}
          toggle={filter.toggleVisibility}
        >
          <PopoverBody>
            {
              {
                [FILTER_TYPES.STRING_FILTER]: <StringFilter {...{ filter }} />,
                [FILTER_TYPES.MULTI_SELECT_FILTER]: (
                  <MultiSelectFilter {...{ filter }} />
                ),
                [FILTER_TYPES.RANGE_FILTER]: <RangeFilter {...{ filter }} />
              }[filter.type]
            }
          </PopoverBody>
        </Popover>
      </span>
    );
  }
}

export default observer(FilterButton);
