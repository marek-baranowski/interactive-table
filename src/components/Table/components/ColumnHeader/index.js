import React from "react";
import { headerTitleStyles } from "./styles";
import SortingButton from "../SortingButton";
import FilterButton from "../FilterButton";

const ColumnHeader = ({ column, sorting, showControls }) => (
  <div className="d-flex">
    {showControls && <SortingButton {...{ column, sorting }} />}
    <span {...headerTitleStyles}>{column.header}</span>
    {showControls && <FilterButton {...{ filter: column.filter }} />}
  </div>
);

export default ColumnHeader;
