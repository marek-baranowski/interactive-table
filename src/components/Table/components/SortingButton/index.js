import React from "react";
import { observer } from "mobx-react";
import { iconStyles } from "../ColumnHeader/styles";
import { SORT_ORDER_TYPES } from "config";

const sortingIconsMapping = {
  [SORT_ORDER_TYPES.ASCENDING]: "sort-asc",
  [SORT_ORDER_TYPES.DESCENDING]: "sort-desc",
  none: "sort"
};

const SortingButton = ({ column: { key, sortable }, sorting }) => {
  if (!sortable) {
    return null;
  }

  const icon =
    sortingIconsMapping[
      sorting.column === key && sorting.order ? sorting.order : "none"
    ];

  return (
    <span
      className={`${iconStyles}`}
      onClick={() => sorting.toggleColumnSorting(key)}
    >
      <i className={`fa fa-${icon}`} />
    </span>
  );
};

export default observer(SortingButton);
