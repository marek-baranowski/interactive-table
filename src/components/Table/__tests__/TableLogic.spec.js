import React from "react";
import { Simulate } from "react-dom/test-utils";
import { Table } from "../index";
import ColumnHeader from "../components/ColumnHeader";
import FilterButton from "../components/FilterButton";
import SortingButton from "../components/SortingButton";
import { FILTER_TYPES, COLUMNS } from "config";
import records from "store/__mockData__/records";
import DataStore from "store/DataStore";

const findHeaderCell = (wrapper, columnKey) => {
  return wrapper.find(ColumnHeader).filterWhere(columnHeader => {
    const { column } = columnHeader.props();

    return !!column && column.key === columnKey;
  });
};

let wrapper;

beforeAll(() => {
  const tableProps = {
    store: DataStore.create({ columns: COLUMNS, records })
  };

  wrapper = mount(<Table {...tableProps} />);
});

test("string filter", () => {
  const filterButton = findHeaderCell(wrapper, "name")
    .find(FilterButton)
    .first()
    .simulate("click");

  const input = filterButton
    .find("Popover")
    .instance()
    ._popover.querySelector("input");

  input.value = "hat";
  Simulate.change(input);

  const filtered = wrapper.prop("store").filteredSortedRecords;
  expect(filtered).toMatchObject([
    { name: "Hattie" },
    { name: "Hattie" },
    { name: "Hatson" }
  ]);
});

test("multi select filter", () => {
  const initialRecordsState = wrapper.prop("store").filteredSortedRecords;

  const filterButton = findHeaderCell(wrapper, "animal")
    .find(FilterButton)
    .first()
    .simulate("click");

  const selectInputs = filterButton
    .find("Popover")
    .instance()
    ._popover.querySelectorAll("[data-test='multi-select'] input");
  expect(selectInputs).toHaveLength(3);

  const [turtle, dog, snake] = selectInputs;

  snake.checked = true;
  Simulate.change(snake);
  expect(wrapper.prop("store").filteredSortedRecords).toMatchSnapshot();

  dog.checked = true;
  Simulate.change(dog);
  expect(wrapper.prop("store").filteredSortedRecords).toMatchSnapshot();

  turtle.checked = true;
  Simulate.change(turtle);
  expect(wrapper.prop("store").filteredSortedRecords).toMatchObject(
    initialRecordsState
  );
});

test("range filter", () => {
  const initialRecordsState = wrapper.prop("store").filteredSortedRecords;

  const rangeFilter = findHeaderCell(wrapper, "price")
    .first()
    .prop("column").filter;

  rangeFilter.setSelectedRange([492, 736]);
  expect(wrapper.prop("store").filteredSortedRecords).toMatchSnapshot();

  rangeFilter.setSelectedRange([285, 911]);
  expect(wrapper.prop("store").filteredSortedRecords).toMatchObject(
    initialRecordsState
  );
});

test("sorting", () => {
  const initialRecordsState = wrapper.prop("store").filteredSortedRecords;

  const sortingButton = findHeaderCell(wrapper, "rating")
    .find(SortingButton)
    .first();

  sortingButton.simulate("click");
  const ascending = wrapper.prop("store").filteredSortedRecords;
  expect(ascending).toMatchSnapshot();

  sortingButton.simulate("click");
  ascending.reverse();
  expect(wrapper.prop("store").filteredSortedRecords).toEqual(ascending);

  sortingButton.simulate("click");
  expect(wrapper.prop("store").filteredSortedRecords).toEqual(
    initialRecordsState
  );
});
