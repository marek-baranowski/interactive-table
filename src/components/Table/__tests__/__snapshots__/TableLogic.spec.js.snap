exports[`test multi select filter 1`] = `
Array [
  Object {
    "animal": "Snake",
    "colour": "White",
    "name": "Hatson",
    "pattern": "Spotted",
    "price": 439,
    "rating": 7,
  },
]
`;

exports[`test multi select filter 2`] = `
Array [
  Object {
    "animal": "Turtle",
    "colour": "Black",
    "name": "Caroline",
    "pattern": "Striped",
    "price": 911,
    "rating": 2,
  },
  Object {
    "animal": "Turtle",
    "colour": "Green",
    "name": "Yvette",
    "pattern": "Striped",
    "price": 653,
    "rating": 4,
  },
  Object {
    "animal": "Turtle",
    "colour": "Black",
    "name": "Hattie",
    "pattern": "Solid",
    "price": 548,
    "rating": 5,
  },
  Object {
    "animal": "Snake",
    "colour": "White",
    "name": "Hatson",
    "pattern": "Spotted",
    "price": 439,
    "rating": 7,
  },
]
`;

exports[`test range filter 1`] = `
Array [
  Object {
    "animal": "Turtle",
    "colour": "Green",
    "name": "Yvette",
    "pattern": "Striped",
    "price": 653,
    "rating": 4,
  },
  Object {
    "animal": "Turtle",
    "colour": "Black",
    "name": "Hattie",
    "pattern": "Solid",
    "price": 548,
    "rating": 5,
  },
]
`;

exports[`test sorting 1`] = `
Array [
  Object {
    "animal": "Turtle",
    "colour": "Black",
    "name": "Caroline",
    "pattern": "Striped",
    "price": 911,
    "rating": 2,
  },
  Object {
    "animal": "Dog",
    "colour": "White",
    "name": "Hattie",
    "pattern": "Solid",
    "price": 738,
    "rating": 3,
  },
  Object {
    "animal": "Turtle",
    "colour": "Green",
    "name": "Yvette",
    "pattern": "Striped",
    "price": 653,
    "rating": 4,
  },
  Object {
    "animal": "Turtle",
    "colour": "Black",
    "name": "Hattie",
    "pattern": "Solid",
    "price": 548,
    "rating": 5,
  },
  Object {
    "animal": "Snake",
    "colour": "White",
    "name": "Hatson",
    "pattern": "Spotted",
    "price": 439,
    "rating": 7,
  },
  Object {
    "animal": "Dog",
    "colour": "White",
    "name": "Guadalupe",
    "pattern": "Spotted",
    "price": 285,
    "rating": 8,
  },
]
`;
